### PR TITLE
New Hero/Code Snippet prototype

### LINKED ISSUE #222: Cards: Add demo with content beside image
Migrated from https://github.com/material-components/material-components-web/issues/4015.

The old catalog had an example of a card containing an image with content beside it. We should add something similar to the new catalog, since there is an example of this type of layout in the [design article](https://material.io/design/components/cards.html#actions).

### PR DESCRIPTION
**Pull request by williamernest**
Implements code snippets in the hero area for button, card, text field, and top app bar.

Also fixes #222.

### CODE REVIEW

File: src/ButtonCatalog.js [hunk 1]
```diff
@@ -91,4 +143,50 @@ class ButtonDemos extends Component {
...
+  options: [
+    {
+      type: 'label',
+      name: 'Options',
+    },
+    {
+      type: 'radiogroup',
+      name: 'Type',
+      urlParam: 'type',
+      value: 'text', // default select first option
```
Comment (moog16) [inline src/ButtonCatalog.js]: can you set this a default in the HeroComponent? So you don't have to set it here

File: src/CodeTemplates.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+export const ReactTemplates = {
+  'ButtonHero': ({label, icon, dense, type, state}) => {
```
Comment (moog16) [inline src/CodeTemplates.js]: can this go into the ButtonComponent as well?

File: src/ComponentCatalogPanel.js [hunk 3]
```diff
@@ -25,15 +26,20 @@ class ComponentCatalogPanel extends Component {
   }
 
   render() {
-    const {designLink, description, demos, docsLink, hero, sourceLink, title} = this.props;
+    const {designLink, description, demos, docsLink, hero, sourceLink, title, config} = this.props;
+    let heroComponent;
+
+    if (config) {
```
Comment (moog16) [inline src/ComponentCatalogPanel.js]: I think add a TODO here saying remove this when we're done with the catalog update

File: src/HeroComponent.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    this.localConfig = JSON.parse(JSON.stringify(this.props.config));
+    const urlParams = queryString.parse(this.props.location.search);
+    this.localConfig = this.copyUrlParamsToLocalConfig(this.localConfig, urlParams);
+  }
+
+  copyUrlParamsToLocalConfig(localConfig, urlParams) {
+
+    // For each url param, copy it over to the local config in the appropriate place.
+    Object.keys(urlParams).forEach((key) => {
+      for (let x = 0; x < localConfig.options.length; x++) {
```
Comment (moog16) [inline src/HeroComponent.js]: can you use a forEach here?

File: src/HeroComponent.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      <WebTab>{this.props.children}</WebTab>,
+      <ReactTab>{this.props.children}</ReactTab>,
+    ];
+    return (
+      <React.Fragment>
+        <TabBar
+            activeIndex={this.state.activeIndex}
+            handleActiveIndexUpdate={(activeIndex) => this.setState({activeIndex})}
+            className='catalog-hero-tab-bar'>
+          <Tab className={'hero-tab'}>
```
Comment (moog16) [inline src/HeroComponent.js]: This section can transform to


```
['Demo', 'Web', 'React'].map((tabName, index) => <Tab className='hero-tab key={index}>
  <span>{tabName}</span>
</Tab>
```

File: src/HeroComponent.js [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    const tabContents = [
+      this.props.children,
+      <WebTab>{this.props.children}</WebTab>,
+      <ReactTab>{this.props.children}</ReactTab>,
+    ];
+    return (
+      <React.Fragment>
+        <TabBar
+            activeIndex={this.state.activeIndex}
+            handleActiveIndexUpdate={(activeIndex) => this.setState({activeIndex})}
```
Comment (moog16) [inline src/HeroComponent.js]: having an anonymous function here causes a second rerender. Do something like
```
handleActiveIndexUpdate={this.handleActiveIndexUpdate}

File: src/HeroComponent.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    this.setState({codeString});
+  };
+
+  render() {
+    return (
+        <React.Fragment>
+          <div style={{display: 'none'}} ref={this.initRef}>{this.props.children}</div>
+          <SyntaxHighlighter
```
Comment (moog16) [inline src/HeroComponent.js]: Would it make sense to pull out common properties between web and react and put it in a constant some where?

File: src/HeroComponent.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+              lineProps={{style: {paddingBottom: 8}}}
+              wrapLines={true}
```
Comment (moog16) [inline src/HeroComponent.js]: also wrapLines should just be:
```
<div wrapLines />
```

File: src/HeroOptionsComponent.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
+import React, {Component} from 'react';
+import {Radio} from './RadioButtonCatalog';
+import {withFormField} from './FormField';
+import queryString from 'query-string';
+import {MDCTextField} from '@material/textfield/index';
+
+export class HeroOptionsComponent extends Component {
+  render() {
+    return (
+        <div className={'hero-options'}>
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: Strings in JSX do not need `{}`

File: src/HeroOptionsComponent.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            })}
+          </ul>
+        </div>
+    );
+  }
+}
+
+const Option = (props) => {
+  return (
+      <React.Fragment>
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: you can just `return getOptionComponent(props.option, props)`

File: src/HeroOptionsComponent.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        {getOptionComponent(props.option, props)}
+      </React.Fragment>
+  )
+};
+
+const getOptionComponent = (option, props) => {
+  if (option.type === 'label') {
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: switch case?

File: src/HeroOptionsComponent.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+  history.push({
+    pathname: history.location.pathname,
+    search: queryString.stringify(urlParams),
+  });
+};
+
+const LabelOption = ({children}) => {
+  return (
+      <span className={'mdc-typography--subtitle2'}>{children}</span>
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: we should get a linter.

File: src/HeroOptionsComponent.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  )
+};
+
+const RadioGroupOption = ({name, options, value, history, urlParams}) => {
+  return (
+      <React.Fragment>
+        <li className={'mdc-list-item'}> <span className={'mdc-typography--subtitle1'}>{name}</span></li>
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: no `{}` around strings

File: src/HeroOptionsComponent.js [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        {options.map((opt) => {
+          return (
+              <li key={opt.value} className={'mdc-list-item'}>
+                <RadioFormField name='demo-radio-set' label={opt.label} defaultChecked={value === opt.value} onChange={() => updateUrl(history, urlParams, 'type', opt.value)}/>
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: change `onChange={this.onChange}`

File: src/HeroOptionsComponent.js [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        })}
+      </React.Fragment>
+  )
+};
+
+class TextFieldOption extends Component {
+  state = {value: ''};
+  constructor(props) {
+    super(props);
+    this.textField = null;
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: delete constructor, and move them to the class.

```
textField = null;
tfRef = React.createRef();
```

File: src/HeroOptionsComponent.js [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+  componentWillUnmount() {
+    if (this.textField) this.textField.destroy();
+  }
+
+
+  render() {
+    return (
+        <React.Fragment>
+          <li key={this.props.name} className={'mdc-list-item'}>
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: no `{}`

File: src/HeroOptionsComponent.js [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+            <div className='mdc-text-field' ref={this.tfRef}>
+              <input type='text' id='my-text-field'
+                     className='mdc-text-field__input' value={this.state.value}
+                     onChange={(event) => {
```
Comment (moog16) [inline src/HeroOptionsComponent.js]: pull onChange out into its own method

File: src/RadioButtonCatalog.js [hunk 18]
```diff
@@ -60,7 +60,8 @@ class Radio extends Component {
                 type='radio'
                 id={this.props.id}
                 name={this.props.name}
-                defaultChecked={this.props.defaultChecked} />
+                defaultChecked={this.props.defaultChecked}
+                onChange={() => this.props.onChange()}/>
```
Comment (moog16) [inline src/RadioButtonCatalog.js]: this should be `onChange={this.props.onChange}`

File: src/ButtonCatalog.js [hunk 19]
```diff
@@ -31,22 +37,68 @@ export class ButtonHero extends Component {
...
+    this.buttonRef = React.createRef();
+  }
+
+  componentWillUnmount() {
+    if (this.ripple) this.ripple.destroy();
+  }
+
+  componentDidUpdate() {
+    if (this.ripple) this.ripple.destroy();
+    this.ripple = new MDCRipple(this.buttonRef.current);
```
[the remark this answers is not in the document]
Comment (williamernest) [reply src/ButtonCatalog.js]: When the component updates to a different style the ripple is broken. Most components will need something like this since we do a lot of setup when initialized.

[the remark this answers is not in the document]
Comment (moog16) [reply src/ButtonCatalog.js]: gotcha

File: src/ButtonCatalog.js (hunk 1 above)
Comment (williamernest) [reply src/ButtonCatalog.js]: Yeah, I could set the default to be the first option if this is null. Is that what you mean?

File: src/CodeTemplates.js (hunk 2 above)
Comment (williamernest) [reply src/CodeTemplates.js]: Yes, I planned to export this from the ButtonCatalog file and import it here so I can export this map. Make sense?

File: src/HeroOptionsComponent.js [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+  componentDidUpdate() {
+    if (this.textField) {
+      this.textField.layout();
+    } else {
+      this.textField = new MDCTextField(this.tfRef.current);
+    }
+  }
+
+  componentDidMount() {
+    this.setState({value: this.props.value});
```
Comment (kfranqueiro) [inline src/HeroOptionsComponent.js]: Is there a reason to do this here rather than doing `this.state = {value: this.props.value}` in constructor?

Comment (williamernest) [reply src/HeroOptionsComponent.js]: We can't call `setState` on a component that isn't mounted.

**Comment by williamernest** [comment]
When react re-renders the button it removes the `mdc-ripple--upgraded` class, but even when there's logic to keep the class on re-render the enhanced ripple stops working the moment the ripple is changed. I'm not sure why it's not working, I'm not seeing any errors being logged. 

Does the Ripple calculate it's size when the user clicks, or is there some kind of initialization going on ? I initially thought it was best to re-instantiate it since the size of the label may change resulting in a smaller/larger ripple.

File: package.json [hunk 21]
```diff
@@ -3,6 +3,9 @@
   "version": "0.1.0",
   "private": true,
   "devDependencies": {
+    "@babel/runtime": "^7.2.0",
```
[the remark this answers is not in the document]
Comment (williamernest) [reply package.json]: react-syntax-highlighter requires it. It fails without it, but I haven't noticed any issues.

File: src/HeroOptionsComponent.js (hunk 20 above)
Comment (kfranqueiro) [reply src/HeroOptionsComponent.js]: I'm not talking about calling setState, I'm talking about assigning it directly. My suggestion comes straight from React's [docs on state & lifecycle](https://reactjs.org/docs/state-and-lifecycle.html#adding-local-state-to-a-class). I'm wondering if there's a specific need to wait until it's mounted, or whether we can actually just initialize the state in the constructor.

**Comment by kfranqueiro** [comment]
> When react re-renders the button it removes the mdc-ripple--upgraded class, but even when there's logic to keep the class on re-render the enhanced ripple stops working the moment the ripple is changed. I'm not sure why it's not working, I'm not seeing any errors being logged.

So like, the ripple just never appears? Are the pseudo-elements getting 0 width/height? My only suspicion is that React might be completely replacing the DOM rather than just re-rendering text and attributes, and you're relying on MDCRipple's vanilla event handlers which were actually hooked up on an element that is no longer present.

> Does the Ripple calculate it's size when the user clicks, or is there some kind of initialization going on ? I initially thought it was best to re-instantiate it since the size of the label may change resulting in a smaller/larger ripple.

Ripple calculates its size when the user clicks. Unbounded ripples additionally call layout the frame after init (and on window resize) so that hover/focus states are sized correctly, but this isn't an unbounded case.

**Comment by williamernest** [comment]
Focus/Hover still work. The event handlers are still the element. The branch now shows what happens if you want to pull it down. Maybe you have an idea since you know more about the ripple than I do.

File: package.json [hunk 22]
```diff
@@ -108,6 +117,6 @@
     "extends": "react-app"
   },
   "dependencies": {
-    "uuid": "^3.3.2"
+    "react-copy-to-clipboard": "^5.0.1"
```
Comment (moog16) [inline package.json]: move to devDeps since its our pattern

File: src/ButtonCatalog.js [hunk 23]
```diff
@@ -16,11 +17,16 @@ const ButtonCatalog = () => {
...
+      initialConfig={ButtonConfig}
+      {...props}
     />
   );
-}
+};
 
-export class ButtonHero extends Component {
+
+// This is retained for the material experiment and will be dead code that we can remove when
```
Comment (moog16) [inline src/ButtonCatalog.js]: whats the status of this?

File: src/ButtonCatalog.js [hunk 24]
```diff
@@ -94,4 +130,76 @@ class ButtonDemos extends Component {
...
+  text: '',
+  raised: 'mdc-button--raised',
+  unelevated: 'mdc-button--unelevated',
+  outlined: 'mdc-button--outlined',
+};
+
+export const ButtonReactTemplate = (config) => {
+  const label = config.options[2].value;
+  const type = config.options[1].value;
+  // TODO: Wire these up when the Config is complete
```
Comment (moog16) [inline src/ButtonCatalog.js]: does this need to be addressed?

File: src/IFrameRoutes.js [hunk 25]
```diff
@@ -21,12 +21,12 @@ import {SliderHero} from './SliderCatalog';
...
 import {SwitchHero} from './SwitchCatalog';
 import {TabsHero} from './TabsCatalog';
-import {TextFieldHero} from './TextFieldCatalog';
+import {TextFieldHeroLegacy} from './TextFieldCatalog';
 import {TopAppBarHero} from './TopAppBarCatalog';
 
 const routesList = [{
   urlPath: 'button',
-  Component: ButtonHero,
+  Component: ButtonHeroLegacy,
```
Comment (moog16) [inline src/IFrameRoutes.js]: do we need a card and top app bar legacy component as well?

File: src/hero/HeroComponent.js [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+              <Tab className='hero-tab' key={index}>
+                <span>{tabName}</span>
+              </Tab>
+            )})}
+        </TabBar>
+        <div className='tab-container'>
+          {tabContents.map((content, index) => {
+            if (this.state.activeIndex !== index) return null;
+
+            ReactGA.event({category: gtagTabCategory, action: 'tab_clicked', label: 'tab_clicked_' + this.tabNames[this.state.activeIndex] });
```
Comment (moog16) [inline src/hero/HeroComponent.js]: this seems like it should be above with <Tab>

```
<Tab onClick={yourCode} className... key={...}```
```

File: src/hero/ReactTab.js [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+  };
+
+  initCodeString = () => {
+    const {children} = this.props;
+    const val = ReactTemplates[children.type.name](children.props.config);
+    const codeString = val ? val.replace(/\n\s*\n/g, '\n') : '';
+    this.setState({codeString});
+  };
+
+  // TODO: Convert this to pass the config object and work for componentDidUpdate
```
Comment (moog16) [inline src/hero/ReactTab.js]: i think we can remove this comment.

File: package.json [hunk 28]
```diff
@@ -35,24 +40,28 @@
     "html-webpack-plugin": "2.29.0",
     "jest": "20.0.4",
     "material-components-web": "^0.43.0",
-    "node-sass": "^4.8.3",
+    "node-sass": "^4.11.0",
```
Comment (kfranqueiro) [inline package.json]: FYI this is going to conflict with master now, since Abhinay switched it to dart-sass. You might want to sync and re-test.

File: src/ButtonCatalog.js [hunk 29]
```diff
@@ -16,11 +17,16 @@ const ButtonCatalog = () => {
       docsLink='https://material.io/components/web/catalog/buttons/'
       sourceLink='https://github.com/material-components/material-components-web/tree/master/packages/mdc-button'
       demos={<ButtonDemos />}
+      initialConfig={ButtonConfig}
+      {...props}
```
Comment (kfranqueiro) [inline src/ButtonCatalog.js]: What is props being used for in the places it's added?

Comment (williamernest) [reply src/ButtonCatalog.js]: It's so that the props are passed into the `ComponentCatalogPanel`. The props have the location information that we're using to store state, so they need to be passed all the way down to the Hero... components.

File: package.json (hunk 28 above)
Comment (moog16) [reply package.json]: Will be handled by merging with master

**Comment by moog16** [comment]
RE ` For some reason the "Properties" heading for Button and "Title" heading for Top App Bar is invisible in Chrome until I select the text. Firefox doesn't have this problem, and it doesn't happen for Text Field in Chrome. This is a weird one.`

Cannot repro - can you help with more information?

**Comment by williamernest** [comment]
Re `* At some point while using the catalog, the drawer and top app bar completely vanished on me. I'm not sure what caused it. I had to refresh to get it back.` 

This is reproducible on the current catalog. I just filed #247 to track it.

File: src/IFrameRoutes.js (hunk 25 above)
Comment (williamernest) [reply src/IFrameRoutes.js]: No, we just need a Button and TextField because we have experiments running for those component.

File: src/ButtonCatalog.js (hunk 23 above)
Comment (williamernest) [reply src/ButtonCatalog.js]: Still required. I'll come back and clean this Legacy classes when the experiment is complete.

**Comment by moog16** [comment]
RE Cards: maybe we can leave a comment in the code that there is some additional setup so that it is obvious. I think the more explicit the better.

RE hero-classes: I think we should be removing. What do you think @williamernest ?

RE top app bar snippets: if we remove the modifier classes like `hero-`, this should avoid this issue, correct?

**Comment by williamernest** [comment]
Re hero classes: I planned on taking a look at that today. We should definitely be removing them

**Comment by moog16** [comment]
RE `mdc-ripple-upgraded` class: not seeing this happen. Which component specifically do you see this? I remember seeing this issue, but can't repro.

File: src/ComponentPage.js [hunk 30]
```diff
@@ -19,23 +21,41 @@ class ComponentPage extends Component {
     this.props.scrollTargetSetter(this.demoContentEl);
   }
 
+  scrollTop = () => {
+    // TODO: when React Tab issue is fixed, this can be removed.
```
Comment (kfranqueiro) [inline src/ComponentPage.js]: Is this irrelevant now? I think we removed React Tab from this PR?

File: src/CheckboxCatalog.js [hunk 31]
```diff
@@ -4,7 +4,7 @@ import {MDCCheckbox} from '@material/checkbox/index';
 
 import './styles/CheckboxCatalog.scss';
 
-const CheckboxCatalog = () => {
+const CheckboxCatalog = (props) => {
```
Comment (kfranqueiro) [inline src/CheckboxCatalog.js]: I assume changes to checkbox/radio/etc are for reuse within the snippet options? Did we try using MDC React components for those?

File: src/IFrameRoutes.js (hunk 25 above)
Comment (kfranqueiro) [reply src/IFrameRoutes.js]: I'm probably missing context here, but where do these routes apply, given that I'm clearly seeing the new stuff when I browse the catalog normally?

File: src/SelectCatalog.js [hunk 32]
```diff
@@ -181,3 +251,4 @@ const Label = ({text, selectId, isEnhanced}) => isEnhanced ? (
 );
 
 export default SelectCatalog;
+export {Select};
```
Comment (kfranqueiro) [inline src/SelectCatalog.js]: Put `export` before class above instead? I think that's what we did in the other instances where we're reusing these?

File: src/hero/CodeTemplates.js [hunk 33]
```diff
@@ -0,0 +1,14 @@
+import {CardReactTemplate} from '../CardCatalog';
+import {TopAppBarReactTemplate} from '../TopAppBarCatalog';
+import {ButtonReactTemplate} from '../ButtonCatalog';
+import {TextFieldReactTemplate} from '../TextFieldCatalog';
+
+const ReactTemplates = {
+  'ButtonHero': (config) => ButtonReactTemplate(config),
+  'CardHero': (config) => CardReactTemplate(config),
+  'CheckboxHero': () => {},
```
Comment (kfranqueiro) [inline src/hero/CodeTemplates.js]: Why is this here?

**Comment by kfranqueiro** [comment]
> RE top app bar snippets: if we remove the modifier classes like `hero-`, this should avoid this issue, correct?

Not sure what you're referring to. My point is the `mdc-top-app-bar--...` modifier classes are being applied to the wrong element entirely because there's a wrapping div around the component.

> RE `mdc-ripple-upgraded` class: not seeing this happen. Which component specifically do you see this? I remember seeing this issue, but can't repro.

Any component that has a ripple applied. Definitely button and card (primary-action).

File: src/CheckboxCatalog.js (hunk 31 above)
Comment (williamernest) [reply src/CheckboxCatalog.js]: We didn't since the components are relatively simple to replicate without any issues.

File: src/IFrameRoutes.js (hunk 25 above)
Comment (williamernest) [reply src/IFrameRoutes.js]: The Legacy routes are used for the experiments for material.io. You can get there by adding `iframe` after component but before button. Once the experiment finishes, we'll delete the Legacy routes and components. 

https://material-components.github.io/material-components-web-catalog/#/component/iframe/button

File: src/ButtonCatalog.js [hunk 34]
```diff
@@ -50,6 +56,36 @@ export class ButtonHero extends Component {
...
+    }
+
+    const className = classnames('hero-button mdc-button', {
+      [ButtonTypes[this.selectedType]]: this.selectedType,
+      'mdc-ripple-upgraded': this.ripple,
+    });
+
+    return (
+      <button className={className} ref={this.initRipple}>
+        {this.props.config.options[2].value}
```
[the remark this answers is not in the document]
Comment (moog16) [reply src/ButtonCatalog.js]: Yes I agree - I'm interested in fixing this. I will fix @williamernest

File: src/hero/CodeTemplates.js (hunk 33 above)
Comment (moog16) [reply src/hero/CodeTemplates.js]: we were also going to group in checkbox to the updated heros. Will remove.

File: src/ButtonCatalog.js (hunk 34 above)
[the remark this answers is not in the document]
Comment (moog16) [reply src/ButtonCatalog.js]: pushed a fix

File: src/ComponentPage.js (hunk 30 above)
Comment (moog16) [reply src/ComponentPage.js]: yup removed!

**Comment by moog16** [comment]
RE top app bar: gotcha - i didn't even notice that since it was working. The only top app bar that didn't work was always short.

RE mdc-ripple-upgraded: still not seeing that :| Maybe try running `npm i`? I can't imagine that would actually fix it. 

Is there anything else outstanding? I think I got all the points besides the mdc-ripple-upgraded.

**Comment by kfranqueiro** [comment]
> RE mdc-ripple-upgraded: still not seeing that :| Maybe try running `npm i`? I can't imagine that would actually fix it.

No, given that I had to run `npm i` to run this branch in the first place.

And yes, I can still reproduce it. I will outline it again, with a specific case:

1. Go to the Button page
2. View the Web tab
3. Change the Variant
4. The Web tab shows `mdc-ripple-upgraded` but shouldn't
5. If you switch to another tab then back, the class will be pruned

File: src/hero/WebTab.js [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+
+  componentDidMount() {
+    this.initCodeString();
+  }
+
+  componentWillUnmount() {
+    this.ripples.map((ripple) => ripple.destroy());
+  }
+
+  componentDidUpdate(prevProps) {
```
Comment (kfranqueiro) [inline src/hero/WebTab.js]: Maybe related to the not-pruning-when-already-in-web-tab bug: I'm noticing this gets called _3 times_ if an option is changed while the web tab is open. It only actually calls `initCodeString` on the 2nd call to this method.

File: src/styles/ComponentCatalogPanel.scss [hunk 36]
```diff
@@ -0,0 +1,85 @@
...
+    margin-bottom: 48px;
+  }
+
+  &__header-elements {
+    max-width: 300px;
+    padding-right: 48px;
+  }
+
+  .hero {
+    order: 4;
```
[the remark this answers is not in the document]
Comment (williamernest) [reply src/styles/ComponentCatalogPanel.scss]: They're used because the mocks show the options before the hero component in desktop but the options come after the hero component in mobile.

File: src/TextFieldCatalog.js [hunk 37]
```diff
@@ -126,29 +130,161 @@ const CharacterCounter = () => (
...
+    initialConfig={getTextFieldConfig(props)}
+    {...props}
   />
 );
 
 this.clickEvent = (label) => () => ReactGA.event({category: gtagCategory, action: gtagTextFieldAction, label: label});
 
-export const TextFieldHero = () => {
+// Keep this until the material io experiment is complete.
+export const TextFieldHeroLegacy = () => {
```
[the remark this answers is not in the document]
Comment (williamernest) [reply src/TextFieldCatalog.js]: `TextFieldHero` and `ButtonHero` no longer need to be exported. They were exported for the iframe routes, but since we've switched to a `...Legacy` export so we don't disturb the experiment in progress.

**Comment by williamernest** [comment]
RE: Height of hero: It all fits above the fold on my 15" MBP with a few hundred pixels to spare (@100% zoom). We did need the extra space before we made leading/trailing icons chips. I can remove 114px from it and it looks better.

File: src/TextFieldCatalog.js [hunk 38]
```diff
@@ -126,29 +130,161 @@ const CharacterCounter = () => (
   <div className='mdc-text-field-character-counter'>0 / {TEXT_FIELD_MAX_LENGTH}</div>
 );
 
-const TextFieldCatalog = () => (
+const getTextFieldConfig = (props) => {
```
[the remark this answers is not in the document]
Comment (moog16) [reply src/TextFieldCatalog.js]: I needed `props` in this component...Other configs do not have multiple values (eg select, text field only have a single value). `icons` is a list of `leading` and/or `trailing`.
